### DISCUSSION TITLE
viewport & scissor's behavior is differ from WebGL

### DISCUSSION THREAD

**Discussion by deepkolos** [gpuweb/gpuweb#4335] [Q&A]
in WebGL

viewport can be any, contained or not, and output is correct

<img width="293" alt="image" src="https://github.com/gpuweb/gpuweb/assets/12824616/fa55d4f9-2f75-418a-84b2-0dc3dac83470">

gl.clear only effect in scissor area

<img width="236" alt="image" src="https://github.com/gpuweb/gpuweb/assets/12824616/4294dce8-1fb5-402f-a9dc-45150cfb65ca">

in WebGPU

viewport must be contained is RT size

clearValue' area is the entire color attachment, ignoring scissor setting

how to early migrate those two behavior from webgl

and here is the use case: use one canvas render all preview result

<img width="946" alt="image" src="https://github.com/gpuweb/gpuweb/assets/12824616/033bc661-5af4-4c17-aec3-8a3ee6d63d04">

WebGL's output

<img width="942" alt="image" src="https://github.com/gpuweb/gpuweb/assets/12824616/b17fc74d-c8a2-40ac-95a5-100b32d5d984">

**Comment by GengineJS** [comment]
Have you solved the problem?

**Reply by jspanchu** [reply]
Is there any update on this? Could the same effect can be achieved by clearing portion of the color attachment - like https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkCmdClearAttachments.html

Is it not implemented in webgpu because others (d3d, metal) don't support rect clears?

**Reply by kainino0x** [reply]
(answer moved below)

**Comment by kainino0x** [comment]
The viewport range issue is #373.

---

About clearing rectangles:

The viewport and scissor intentionally don't apply to render pass load/store ops, and this is based on the designs of vulkan/metal (and d3d12 after it added render passes). The viewport and scissor state is inside the render pass, but the load/store ops apply to the whole pass, so they can't use any state from inside the pass.

In order to clear a rectangle you unfortunately need to draw a quad, with a solid color. It can be a fullscreen quad, because you can use viewport or scissor to constrain the quad to the region you want to draw.

**Reply by jspanchu** [reply]
Got it. The opengl code I'm trying to port relied upon constrained color/depth clears based on viewport parameters. For example, a side-by-side viewport configuration in our engine roughly boiled down these oversimplified opengl calls:

```c
// The left half has a specific background color r1, g1, b1, a1 and draws some triangles.
glViewport(0, 0, 0.5, 1.0)
glClearColor(r1, g1, b1, a1); glClear(clear_bits)
glDrawArrays(...)
// The right half has a specific background color r2, g2, b2, a2 and draws some triangles.
glViewport(0.5, 0, 0.5, 1.0)
glClearColor(r2, g2, b2, a2); glClear(clear_bits)
glDrawArrays(...)
```

> In order to clear a rectangle you unfortunately need to draw a quad, with a solid color. It can be a fullscreen quad, because you can use viewport or scissor to constrain the quad to the region you want to draw.

Okay, looks like we'll have to change the implementation of backgrounds in our engine to draw a scissored quad, and then draw the meshes on top of that using loadOp = `Load` instead of `Clear` for the color attachment.

**Reply by kainino0x** [reply]
BTW one small trick to make it easier to draw a solid-color quad is to use
`blend: { color: { srcFactor: "constant" }, alpha: { srcFactor: "constant" } }`
in the render pipeline, then you can use `setBlendConstant()` to configure the clear color inline in the render pass.
